File: backend/app/schemas/onboarding.py
"""Onboarding schemas for the healthcare IVR platform."""

from datetime import datetime
from typing import Optional, Dict, Any, List
from pydantic import BaseModel, Field
from uuid import UUID


class OnboardingStepBase(BaseModel):
    """Base schema for onboarding steps."""
    step_name: str = Field(..., description="Name of the onboarding step")
    step_order: int = Field(
        ..., description="Order of the step in the sequence"
    )
    completed: bool = Field(
        default=False, description="Whether step is completed"
    )
    data: Optional[Dict[str, Any]] = Field(
        default_factory=dict, description="Additional step data"
    )


class OnboardingStepCreate(OnboardingStepBase):
    """Schema for creating onboarding steps."""
    user_id: UUID = Field(..., description="User ID for the onboarding step")


class OnboardingStepUpdate(BaseModel):
    """Schema for updating onboarding steps."""
    completed: Optional[bool] = Field(
        None, description="Whether step is completed"
    )
    data: Optional[Dict[str, Any]] = Field(
        None, description="Additional step data"
    )


class OnboardingStepResponse(OnboardingStepBase):
    """Schema for onboarding step responses."""
    id: UUID = Field(..., description="Unique identifier for the step")
    user_id: UUID = Field(..., description="User ID for the onboarding step")
    completed_at: Optional[datetime] = Field(None, description="When step was completed")
    created_at: datetime = Field(..., description="When step was created")
    updated_at: datetime = Field(..., description="When step was last updated")

    class Config:
        from_attributes = True


class OnboardingProgressResponse(BaseModel):
    """Schema for overall onboarding progress."""
    user_id: UUID = Field(..., description="User ID")
    total_steps: int = Field(..., description="Total number of onboarding steps")
    completed_steps: int = Field(..., description="Number of completed steps")
    current_step: Optional[str] = Field(None, description="Current step name")
    progress_percentage: float = Field(..., description="Progress as percentage")
    onboarding_completed: bool = Field(..., description="Whether onboarding is complete")
    onboarding_started_at: Optional[datetime] = Field(None, description="When onboarding started")
    onboarding_completed_at: Optional[datetime] = Field(None, description="When onboarding completed")
    steps: List[OnboardingStepResponse] = Field(..., description="List of onboarding steps")

    class Config:
        from_attributes = True


class OnboardingStartRequest(BaseModel):
    """Schema for starting onboarding."""
    skip_welcome: bool = Field(default=False, description="Whether to skip welcome step")


class OnboardingCompleteRequest(BaseModel):
    """Schema for completing onboarding."""
    feedback: Optional[str] = Field(None, description="User feedback about onboarding")
    rating: Optional[int] = Field(None, ge=1, le=5, description="Rating from 1-5")


class RoleOnboardingConfig(BaseModel):
    """Schema for role-specific onboarding configuration."""
    role: str = Field(..., description="User role")
    steps: List[Dict[str, Any]] = Field(..., description="Onboarding steps configuration")
    estimated_duration: int = Field(..., description="Estimated duration in minutes")
    welcome_message: str = Field(..., description="Welcome message for the role")
    completion_message: str = Field(..., description="Completion message for the role")


class OnboardingAnalytics(BaseModel):
    """Schema for onboarding analytics."""
    total_users: int = Field(..., description="Total users with onboarding")
    completed_users: int = Field(..., description="Users who completed onboarding")
    completion_rate: float = Field(..., description="Completion rate as percentage")
    average_completion_time: Optional[float] = Field(None, description="Average completion time in hours")
    step_completion_rates: Dict[str, float] = Field(..., description="Completion rates by step")
    role_completion_rates: Dict[str, float] = Field(..., description="Completion rates by role")

    class Config:
        from_attributes = True

File: frontend/src/types/onboarding.ts
/**
 * Onboarding types for the healthcare IVR platform.
 */

export interface OnboardingStep {
  id: string;
  user_id: string;
  step_name: string;
  step_order: number;
  completed: boolean;
  completed_at?: string;
  data: Record<string, any>;
  created_at: string;
  updated_at: string;
}

export interface OnboardingProgress {
  user_id: string;
  total_steps: number;
  completed_steps: number;
  current_step?: string;
  progress_percentage: number;
  onboarding_completed: boolean;
  onboarding_started_at?: string;
  onboarding_completed_at?: string;
  steps: OnboardingStep[];
}

export interface OnboardingStepUpdate {
  completed?: boolean;
  data?: Record<string, any>;
}

export interface OnboardingStartRequest {
  skip_welcome?: boolean;
}

export interface OnboardingCompleteRequest {
  feedback?: string;
  rating?: number;
}

export interface RoleOnboardingConfig {
  role: string;
  steps: OnboardingStepConfig[];
  estimated_duration: number;
  welcome_message: string;
  completion_message: string;
}

export interface OnboardingStepConfig {
  name: string;
  title: string;
  description: string;
  duration: number;
  icon?: string;
  instructions?: string;
}

export interface OnboardingAnalytics {
  total_users: number;
  completed_users: number;
  completion_rate: number;
  average_completion_time?: number;
  step_completion_rates: Record<string, number>;
  role_completion_rates: Record<string, number>;
}

export type OnboardingStepName =
  | 'welcome'
  | 'profile_setup'
  | 'patient_management'
  | 'ivr_workflow'
  | 'dashboard_tour'
  | 'review_queue'
  | 'approval_workflow'
  | 'communication_tools'
  | 'doctor_management'
  | 'schedule_setup'
  | 'analytics_overview'
  | 'order_management'
  | 'shipping_logistics'
  | 'analytics_reports'
  | 'order_queue'
  | 'shipment_tracking'
  | 'territory_management'
  | 'user_management'
  | 'system_configuration'
  | 'audit_compliance'
  | 'program_management'
  | 'community_partners'
  | 'compliance_tracking'
  | 'shipping_queue'
  | 'carrier_management'
  | 'warehouse_config';

export type UserRole =
  | 'Doctor'
  | 'IVR'
  | 'Sales'
  | 'Master Distributor'
  | 'Distributor'
  | 'Admin'
  | 'CHP Admin'
  | 'Shipping and Logistics';